File: python_tests/reuse_distance.py
"""

Prompt: Implement a python program for me using the Stack based Naive Method described in the following document https://www.cc.gatech.edu/classes/AY2019/cs7260_spring/lecture30.pdf#:~:text=Reuse%20distance%20is%20the%20number%20of%20distinct%20memory,was%20the%20last%20time%20you%20used%20this%20address.

This program includes a `main` method that creates an instance of the `ReuseDistance` class and processes a list of addresses using the `process_address` method. At the end, it calls the `get_reuse_distance` method to print the final reuse distance.
Is there anything else you would like to add to this program?

Source: Conversation with Bing, 5/2/2023(1) Data Structures and Algorithms for Calculating Reuse Distance. https://www.cc.gatech.edu/classes/AY2019/cs7260_spring/lecture30.pdf Accessed 5/2/2023.
(2) CS7260 Spring 2017. https://www.cc.gatech.edu/classes/AY2019/cs7260_spring/ Accessed 5/2/2023.
(3) Schedule of Classes | Registrar's Office | Georgia Tech. https://registrar.gatech.edu/current-students/schedule-of-classes Accessed 5/2/2023.

"""

class ReuseDistance:
    def __init__(self):
        self.stack = []
        self.reuse_distance = []

    def process_address(self, address):
        if address not in self.stack:
            self.stack.append(address)
            self.reuse_distance.append(float('inf'))
        else:
            temp_stack = []
            while self.stack[-1] != address:
                temp_stack.append(self.stack.pop())
            self.stack.pop()
            reuse_distance = len(temp_stack) + 1
            self.reuse_distance.append(reuse_distance)
            while temp_stack:
                self.stack.append(temp_stack.pop())
            self.stack.append(address)

    def get_reuse_distance(self):
        return self.reuse_distance

def main():
    addresses = ['a', 'b', 'b', 'c', 'd', 'a', 'e', 'f', 'g']
    rd = ReuseDistance()
    for address in addresses:
        rd.process_address(address)
    print(rd.get_reuse_distance())

if __name__ == '__main__':
    main()

File: python_tests/reuse_distance.cpp
/*
Sure! Here's a C++ program that implements the Stack based Naive Method for calculating reuse distance as described in the document you provided earlier:

This program implements a class called `ReuseDistance` that uses a stack to calculate reuse distance. The `process_address` method takes an address as input and updates the stack and reuse distance accordingly. The `get_reuse_distance` method returns the current reuse distance.

The `main` function creates an instance of the `ReuseDistance` class and processes a list of addresses using the `process_address` method. At the end, it calls the `get_reuse_distance` method to print the final reuse distance.
*/

#include <iostream>
#include <stack>
#include <vector>
#include <algorithm>

class ReuseDistance {
    private:
        std::stack<char> stack;
        std::vector<int> reuse_distance;
    public:
        void process_address(char address) {
            std::stack<char> temp_stack;
            bool found = false;
            while (!stack.empty()) {
                if (stack.top() == address) {
                    found = true;
                    stack.pop();
                    break;
                }
                temp_stack.push(stack.top());
                stack.pop();
            }
            if (!found) {
                reuse_distance.push_back(INT_MAX);
            } else {
                int rd = temp_stack.size() + 1;
                reuse_distance.push_back(rd);
            }
            while (!temp_stack.empty()) {
                stack.push(temp_stack.top());
                temp_stack.pop();
            }
            stack.push(address);
        }

        std::vector<int> get_reuse_distance() {
            return reuse_distance;
        }
};

int main() {
    std::vector<char> addresses = {'a', 'b', 'b', 'c', 'd', 'a', 'e', 'f', 'g'};
    ReuseDistance rd;
    for (char address : addresses) {
        rd.process_address(address);
    }
    std::vector<int> reuse_distance = rd.get_reuse_distance();
    for (int rd : reuse_distance) {
        std::cout << rd << " ";
    }
    std::cout << std::endl;
    return 0;
}


